schema-layer/data-structures/hashmap.md

```diff
@@ -0,0 +1,298 @@
+
+Insertion of new buckets with `Set(key, value)` involves splicing in a new element to the data array at the `dataIndex` position and setting the `index` bit of the `map` bitmap. Converting a bucket to a child node leaves the `map` bit map alone as the `index` bit still indicates there is an element at that position.
+
+A `Get(key)` operation performs the same hash, `index` and `dataIndex` calculation at the root node, traversing into a bucket to find an entry matching `key` or traversing into child nodes and performing the same `index` and `dataIndex` calculation but at an offset of an additional `bitWidth` bits in the `key`'s hash.
+
+A `Delete(key)` mutation first locates the element in the same way as `Get(key)` and if that entry exists, it is removed from the bucket containing it. If the bucket is empty after deletion of the entry, we remove the bucket element completely from the data element array and unsets the `index` bit of `map`. If the node containing the deleted element has no links to child nodes and contains `bucketSize` elements after the deletion, those elements are compacted into a single bucket and placed in the parent node in place of the link to that node. We perform this check on the parent (and recursively if required), thereby transforming the tree into its most compact form, with only buckets in place of nodes that have up to `bucketSize` entries at all edges. This compaction process combined with the `key` ordering of entries in buckets produces canonical forms of the data structure for any given set of `key` / `value` pairs regardless of their insertion order or whether any intermediate entries have been added and deleted.
+
+By default, each node in an IPLD HashMap is stored in a distinct IPLD block and CIDs are used for child node links. The schema and algorithm presented here also allows for inline child nodes rather than links, with read operations able to traverse multiple nodes within a single block where they are inlined. The production of inlined IPLD HashMaps is left unspecified and users should be aware that inlining breaks canonical form guarantees.
+
+## Structure
+
+### Parameters
+
+Configurable parameters for any given IPLD HashMap:
```

Comment:
I would really like to see these being optional, such that we don't have to add the overhead of storing these in every hamt that gets serialized (especially for smaller structures)

Reply:
bonus: if these become optional a regular node and a root node could be the same data structure, allowing for some interesting nesting

Reply:
@dignifiedquire: optional means we're going to need to be more confident of the default values and their future suitability. I don't think we have that, we don't know how these data structures will be used by real-world applications or what the implications of the default shapes we're defining here are. For example, I have a suspicion that buckets might not be the right approach and that the elasticity in block sizes they will yield might be too messy. But I can't make assertions on that without real world applications and a lot of research work.

My preference would be to stick with explicit definitions for now, and maybe over time it'll be clear what the right parameters are an define implicit defaults.

> allowing for some interesting nesting

Not really though, depth changes everything, a depth=0 node (root) is useless at depth=>0. Do you have something concrete in mind?

Reply:
> Do you have something concrete in mind?

no, just a random idea

> My preference would be to stick with explicit definitions for now

fair enough

Reply:
I don't even think we need to put them in there. We already have to have context to know that this is a HAMT and how to traverse it. might as well say that we need context to know the width.

Additionally, we can infer the width from the bitfield.

Reply:
Well, that goes back to the discussion @ https://github.com/ipld/specs/issues/130 I think so I'd rather that all be resolved before I remove these. Right now they're useful since we have no formal way of defining "context" except by inlining it. Fat pointers? CIDv2? We have neither.

The current experiment that I'm working with, the only thing that currently leverages these advanced layouts / generics / composites, is @mikeal's js-types which embeds a `_type` property that points to the implementation. Search for `_type` in https://github.com/ipld/js-generics/pull/1/files to see it embedded in blocks as they are encoded for a HashMap. The js-types implementation then knows when it encounters a `_type` at the root of a block that it can go and find the implementation that matches that string and traverse the advanced layout / generic / composite with the associated logic.

I understand this may not be ideal, but it works at least and in lieu of an alternative I'm not sure how to make this useful without hardwiring all of these parameters into code that creates and reads these blocks.

Good point re `width`, I'll try removing that from my implementation and update here.

Reply:
So, we shouldn’t need an identifier like `_type` in *this* spec. For the (advanced layouts, / generics / composites) I’m separating out the “definition,” which describes the implementation of the data structure, from the actual data. Everything in this spec, including these parameters, are part of the data.

It’s important to keep these separate because we’re exploring in-band **and** out-of-band signaling mechanisms. Since the “definition” can be applied in multiple ways we need to make sure that definitions stay out of specs like this. As @whyrusleeping notes, there are cases where you already know what data structure you’re working with and don’t even need an in-band description.

The experimental implementation we have will change once we figure out how the “fat pointer” works and will only read the `_type` field when we encounter the right kind of link *or* when you explicitly tell it the data you’re passing is a definition.

Reply:
I'll admit that I don't fully understand this desire to remove this kind of data from blocks. I've been imagining that making blocks as self describing as possible was a worthy goal. There are obvious limitations to that but CIDs seem to get us over most of those. It seems to me that once you have a CID for _something_ that's all you ideally should need to understand what's in it—minus all of the programatic pieces to decode and navigate through it (codecs and these advanced layout pieces). There seems to be a desire to push a lot of that self describing stuff into the "how we got to here" place. That's the bit I don't get.

Reply:
This got a little off-track, let me try and float a proposal that I think can resolve it.

In our experimental composite system, we could have something like this:

```javascript
{ _type: ‘IPLD/Experimental/HAMT/0’,
  hamt: { 
    hashAlg: ‘sha2-256’,
    bitWidth: 8,
    bucketSize: 3,
    root: {
       data: [],
       map: Bytes(),
    }
  }
}
```  

The top level node containing `_type` and `hamt` are part of the “definition” which isn’t part of this spec and is only there for illustrative purposes.

The root block of this spec is then modified to put `bytes` and `map` under a `root` key. In the schema we can define this as a union of Map or Link. In the Filecoin use case, they’d just stick to encoding the root node all the way down. For everyone else, including people linking to Filecoin’s hamt, they can create a block containing the parameters and then link to this root block.

This gives us the flexibility to avoid encoding parameters which are constant in the application while still allowing people to create pointers to the data containing the parameter context they need in order to read the structure properly. The only thing we need to make sure to note is that if you inline the Map vs Link to a Map you’re going to end up with a different hash.